feat(standups): Standup UI - last updated time #6393

JimmyLv opened the pull request:
implements #6393

based on #6469 (need to merge firstly)

![Google Chrome-Async Standup #2  Jimmy Lv’s Team-002274-20220513](https://user-images.githubusercontent.com/4997466/168316878-c53e7d5f-f0da-4293-bee0-45d6bb5765ce.png)

JimmyLv commented on packages/client/components/TeamPrompt/TeamPromptResponseCard.tsx:
```diff
@@ -123,15 +133,14 @@ const TeamPromptResponseCard = (props: Props) => {
   const {teamMember, meetingId, meeting, discussion, response} = responseStage
   const {picture, preferredName, userId} = teamMember
 
+  const discussionEdges = discussion.thread.edges
+  const replyCount = discussionEdges.length
+
```
just moved the position because of a merge conflict. Can be ignored

JimmyLv commented:
@jmtaber129 Hi James, could you please help to review this PR? 😁

jmtaber129 commented on packages/client/components/TeamPrompt/TeamPromptLastUpdatedTime.tsx:
```diff
@@ -0,0 +1,26 @@
+  const formattedTime = updatedAt ? relativeDate(updatedAt) : relativeDate(createdAt)
+  return (
+    <Timestamp>
+      <span>{formattedTime}</span>
+      {isEdited && <span> · Edited</span>}
```
+1 No need for `span` + simplify to one block:
```suggestion
      {formattedTime}{isEdited && ' · Edited'}
```

jmtaber129 commented on packages/client/components/TeamPrompt/TeamPromptResponseCard.tsx:
```diff
@@ -81,11 +82,20 @@ const TeamPromptResponseCard = (props: Props) => {
           picture
           preferredName
         }
+        discussion {
+          thread(first: 1000) @connection(key: "DiscussionThread_thread") {
+            edges {
+              ...TeamPromptRepliesAvatarList_edges
+            }
+          }
+        }
```
+1 duplicates the below `discussion`

jmtaber129 commented on packages/client/mutations/UpsertTeamPromptResponseMutation.ts:
```diff
@@ -13,6 +13,7 @@ graphql`
       userId
       content
       plaintextContent
+      updatedAt
```
-1 this needs `createdAt` for new responses

jmtaber129 commented on packages/client/mutations/UpsertTeamPromptResponseMutation.ts:
```diff
@@ -77,7 +79,8 @@ const UpsertTeamPromptResponseMutation: StandardMutation<
         content,
-        plaintextContent
+        plaintextContent,
+        updatedAt: now
```
-1 this needs `createdAt` for new responses

jmtaber129 commented on packages/client/components/TeamPrompt/TeamPromptLastUpdatedTime.tsx:
```diff
@@ -0,0 +1,26 @@
+
+export default function LastUpdatedTime({updatedAt, createdAt}: Props) {
+  const isEdited = createdAt !== updatedAt
+  const formattedTime = updatedAt ? relativeDate(updatedAt) : relativeDate(createdAt)
```
+1 `updatedAt` should always be non-null, so I think this can be simplified a bit

jmtaber129 reviewed (changes requested):
Looking good!

A few high-level comments:
1. -1 After editing, the timestamp flashes to the optimistic update and then immediately back to the old timestamp.  This seems to be because we're not clearing the cache in the upsert mutation (this hasn't been an issue since we're manually setting the content, but that seems like a smell).  Applying this patch should fix this, and prevent similar issues from popping up in the future:
```
diff --git a/packages/server/graphql/public/mutations/upsertTeamPromptResponse.ts b/packages/server/graphql/public/mutations/upsertTeamPromptResponse.ts
index 4473e5505..3842bb612 100644
--- a/packages/server/graphql/public/mutations/upsertTeamPromptResponse.ts
+++ b/packages/server/graphql/public/mutations/upsertTeamPromptResponse.ts
@@ -67,11 +67,10 @@ const upsertTeamPromptResponse: MutationResolvers['upsertTeamPromptResponse'] =
     content,
     plaintextContent
   })
-  const updatedTeamPromptResponse = await dataLoader
+
+  dataLoader
     .get('teamPromptResponses')
-    .load(teamPromptResponseId)
-  updatedTeamPromptResponse.content = contentJSON
-  updatedTeamPromptResponse.plaintextContent = plaintextContent
+    .clear(teamPromptResponseId)

   const data = {
     meetingId,
```
2. -1 Upon creating a new response, the timestamp doesn't update to reflect the `createdAt` time until after refresh (see inline comments for where `createdAt` is missing)
3. -1 A pattern I've seen in other apps like Slack is to use the original message timestamp (`createdAt` in this case) for the primary timestamp, and then surface the edit time through a tooltip over `Edited`.  I think this is the way to go here, but I'd get some design input from @ackernaut 
4. +1 The timestamp introduces some inconsistency with spacing + card size between the header and the actual card for empty cards vs. responded cards (see screenshot) - a card with the timestamp actually more closely matches the [designs](https://www.figma.com/file/bhiEOV4HK9Kv91HFsJmevV/Stand-up-Concepts?node-id=1147%3A73406), but I think we need to enforce a 12px bottom margin on the avatar component to ensure consistency
<img width="682" alt="Screen Shot 2022-05-16 at 1 48 49 PM" src="https://user-images.githubusercontent.com/9013217/168662630-eb280b31-9274-4595-a7fb-41d779e979b4.png">
5. +1 A tooltip over the relative timestamp showing the actual timestamp might be nice.  @ackernaut what do you think?

ackernaut commented:
@jmtaber129 
- So simply show ‘Edited’ and when you hover there’s a tooltip to show the timestamp for when it was edited? That sounds straightforward enough to me 👍 
- When the relative timestamp says something like ‘an hour ago’ you can hover that and the timestamp gives you something like ‘Mon May 16 14:34’? I like it 👍

jmtaber129 commented:
@ackernaut yep!  Just to clarify, overall what this would look like is we'd have the format `<createdAt><' * Edited' if edited>`, where hovering over `createdAt` would show a timestamp like `Mon May 16 14:34`, and hovering over `Edited` would show the `updatedAt` timestamp

ackernaut commented:
@jmtaber129 LGTM!

JimmyLv commented:
wow, that's so nice, thanks @jmtaber129 for reviewing timely, let me update!

JimmyLv commented:
@jmtaber129 Hey James, I've fixed all your feedback and updated the tooltip interaction behaviors, can you help review them again? Thanks!

![Google Chrome-Async Standup #3  Jimmy Lv’s Team-002439-20220520](https://user-images.githubusercontent.com/4997466/169491011-138acad9-4241-4218-ba1b-8993fa01881e.png)

JimmyLv commented:
@jmtaber129 Hey James, I've fixed all your feedback and updated the tooltip interaction behaviors, can you help review them again? Thanks!

![Google Chrome-Async Standup #3  Jimmy Lv’s Team-002439-20220520](https://user-images.githubusercontent.com/4997466/169491011-138acad9-4241-4218-ba1b-8993fa01881e.png)

jmtaber129 commented on packages/client/mutations/UpsertTeamPromptResponseMutation.ts:
```diff
@@ -77,7 +80,9 @@ const UpsertTeamPromptResponseMutation: StandardMutation<
-        plaintextContent
+        plaintextContent,
+        updatedAt: now,
+        createdAt: now
```
+1 This causes the createdAt time to flash from `X minutes ago` to `just now` and back.  I think we should only provide this if a response doesn't already exist (i.e. when `teamPromptResponseId` is falsey).  That way, we optimistically provide a `createdAt` of `now` for new responses, and let relay use the value already in the store for updated responses.

```suggestion
        createdAt: !teamPromptResponseId ? now : undefined
```

JimmyLv replied:
Yes, that's the case!

jmtaber129 reviewed (approved):
Looking good!  A few minor issues (two inline comment + the below comment)

+1 `createdAt` still doesn't show up until after refresh, but I found the fix (unrelated to this PR, there's an updater that we need for newly inserted responses - new responses don't get upserted into the local store because the ID doesn't already exist, so filling out an initial response, navigating away from the meeting, and then navigating back to the meeting doesn't show the new response).  Try applying this patch and it should work:
```
diff --git a/packages/client/mutations/UpsertTeamPromptResponseMutation.ts b/packages/client/mutations/UpsertTeamPromptResponseMutation.ts
index f13ea9d9c..2d0b09e33 100644
--- a/packages/client/mutations/UpsertTeamPromptResponseMutation.ts
+++ b/packages/client/mutations/UpsertTeamPromptResponseMutation.ts
@@ -91,6 +91,10 @@ const UpsertTeamPromptResponseMutation: StandardMutation<
     mutation,
     variables,
     optimisticResponse,
+    updater:  (store) => {
+      const payload = store.getRootField('upsertTeamPromptResponse')
+      upsertTeamPromptResponseUpdater(payload as any, {atmosphere, store})
+    },
     onCompleted,
     onError
   })
```

cc @Dschoordsch @BartoszJarocki since this was removed as a result of review of https://github.com/ParabolInc/parabol/pull/6504.  Definitely a weird "gotcha".

JimmyLv commented:
@Dschoordsch Hey Georg, could you please help to do the maintainer review for this PR? I'd like to ensure this one can be merged into the next release, thanks! 😁

Dschoordsch commented:
@JimmyLv please add me to the reviewers once I should do a review, otherwise I might miss it.

Dschoordsch commented on packages/client/components/promptResponse/PromptResponseEditor.tsx:
```diff
@@ -56,7 +57,11 @@ const PromptResponseEditor = (props: Props) => {
-    handleSubmit && handleSubmit(newEditorState)
+    const newContent = newEditorState.getJSON()
+
+    if (areEqual(content, newContent)) return
```
-1 I think you need to check here for empty and undefined/null to avoid creating an empty post on first blur.

JimmyLv replied:
done! good point, and we can use `newEditorState.isEmpty` to check the empty input content

Dschoordsch reviewed (changes requested):
It feels weird that the first blur of the input field creates a response, even when it is empty.
![Updated at is set when loosing focus](https://user-images.githubusercontent.com/7331043/170114349-badac57e-6cca-4727-9a3e-758e517aa9b4.gif)

I also would like to have the times updated, but that could happen separately. If you don't address this, make sure to create a ticket for it.

JimmyLv commented on packages/client/components/TeamPrompt/TeamPromptLastUpdatedTime.tsx:
```diff
@@ -0,0 +1,52 @@
+  const isEdited = createdAt !== updatedAt
+  return (
+    <Timestamp>
+      <Hover onMouseEnter={showCreatedTime} onMouseLeave={closeCreatedTime} ref={createdTimeRef}>
```
I see what you mean, it's because the hover triggers a re-render of the component, so there's a jump from "just now" to "3 minutes ago", but do you expect this time to keep updating automatically? That might require a poll, is that too heavy or do you have other suggestions for it?

Dschoordsch replied:
We don't need to poll the server, we would just need to reformat the time. Ideally this would be done outside of React component tree by updating the ref via script directly, i.e. add a setInterval which takes the ref of the text component and updates it without doing a full re-render. @nickoferrall I think you have more React experience, what do you think?

nickoferrall replied:
Yep, I'd add a setInterval to a useEffect and update the timestamp that way. Interestingly, in our Discussion Thread, I noticed we're not updating the timestamp until the component rerenders, so if you implement a pattern in standups, it could be useful elsewhere in the app.

JimmyLv commented:
Hey @Dschoordsch , I updated the first empty check to avoid creating an empty post on the first blur. Could you please help to review again, thanks!

https://user-images.githubusercontent.com/4997466/170245563-5e4453f5-2a68-4aae-928f-2cc093b3f1ca.mp4
